addons/survey/tests/test_survey.py

```diff
@@ -296,3 +297,44 @@ def get_question_by_title(survey, title):
         self.assertNotEqual(q_food_not_vegetarian_cloned.triggering_question_id.id, q_is_vegetarian.id)
         self.assertNotEqual(q_food_not_vegetarian_cloned.triggering_answer_id.id,
                             q_is_vegetarian.suggested_answer_ids[1].id)
+
+    def test_get_questions_except_invalid_conditionals(self):
+        """ Tests the method `_get_questions_except_invalid_conditionals`. This method takes a
+            recordset of question.question from a survey.survey and returns a recordset without
+            conditional questions where the condition is not valid. """
+
+        survey_feedback = self.env.ref('survey.survey_feedback')
+        passed_questions = survey_feedback.question_and_page_ids.filtered(
+            lambda question: not question.is_page or not is_html_empty(question.description)).sorted()
```

Comment:
Why `.sorted()`?

Reply:
It's true it is not necessary here. However, is it ok to use it after changing the sequence? I've noticed that the recordsets passed to `_get_pages_or_questions_except_invalid_conditionals` are sorted in the code but were not in my tests, which messed up the result. I don't know where the "magic" of the sorting happens however

Reply:
I see the problem.
No part of the code should rely on the order of records in a 'base' recordset (more below). 
As any "set", you should see them as bags, stuff inside are free to move around at will.
It means that even if you sort it, it could not stay like this afterwards.
(There is an OrderedSet class in odoo too for special cases that do need to keep the order, but I don't think it's required here.)
You can however use sorting when iterating its values, where you **need** to go through them in a specific order, and this is the case here as we build the invalid set as we go through the loop.
https://github.com/odoo/odoo/blob/393787b10e81a4b09345dc78591b8ef7a8941bc2/addons/survey/models/survey_survey.py#L596-L598
(/spoiler: other hint: make sure to specify how sorting should be done.)

Reply:
Is it not ok to not specify a sorting, as the `sorted` function with no parameters sorts according to the model's order - here it is what I need (`sequence, id`) and I expect it won't change in the future?

Reply:
Oh, indeed you don't need to specify the sorting method (and shouldn't :+1: , bad hint :confused:  ) !